File: js/lib/CodebookName.ts
export enum CodebookName {
    Quantity = 1,
    Content = 2,
    Calculation = 3,
    State = 4,
    Command = 5,
    Type = 6,
    FunctionalServices = 7,
    MaintenanceCategory = 8,
    ActivityType = 9,
    Position = 10,
    Detail = 11,
}

export class CodebookNames {
    public static get names(): CodebookName[] {
        return Object.keys(CodebookName)
            .filter((v) => !isNaN(+v))
            .map((v) => +v as CodebookName);
    }

    public static fromString(name?: string): CodebookName {
        switch (name?.toLocaleLowerCase()) {
            case "position":
            case "positions":
                return CodebookName.Position;
            case "calculation":
            case "calculations":
                return CodebookName.Calculation;
            case "quantity":
            case "quantities":
                return CodebookName.Quantity;
            case "state":
            case "states":
                return CodebookName.State;
            case "content":
            case "contents":
                return CodebookName.Content;
            case "command":
            case "commands":
                return CodebookName.Command;
            case "type":
            case "types":
                return CodebookName.Type;
            case "detail":
                return CodebookName.Detail;
            case "functionalService":
                return CodebookName.FunctionalServices;
            case "maintenanceCategories":
                return CodebookName.MaintenanceCategory;
            case "activityType":
                return CodebookName.ActivityType;
            default:
                throw new Error(`Unknown or invalid metadata tag: ${name}`);
        }
    }

    public static fromPrefix(prefix?: string): CodebookName {
        switch (prefix) {
            case "pos":
                return CodebookName.Position;
            case "calc":
                return CodebookName.Calculation;
            case "qty":
                return CodebookName.Quantity;
            case "state":
                return CodebookName.State;
            case "cnt":
                return CodebookName.Content;
            case "cmd":
                return CodebookName.Command;
            case "type":
                return CodebookName.Type;
            case "detail":
                return CodebookName.Detail;
            case "funct.svc":
                return CodebookName.FunctionalServices;
            case "maint.cat":
                return CodebookName.MaintenanceCategory;
            case "act.type":
                return CodebookName.ActivityType;

            default:
                throw new Error(`Unknown or invalid prefix: ${prefix}`);
        }
    }

    public static toString(name: CodebookName): string {
        switch (name) {
            case CodebookName.Position:
                return "position";
            case CodebookName.Calculation:
                return "calculation";
            case CodebookName.Quantity:
                return "quantity";
            case CodebookName.State:
                return "state";
            case CodebookName.Content:
                return "content";
            case CodebookName.Command:
                return "command";
            case CodebookName.Type:
                return "type";
            case CodebookName.Detail:
                return "detail";
            case CodebookName.FunctionalServices:
                return "functionalService";
            case CodebookName.MaintenanceCategory:
                return "maintenanceCategory";
            case CodebookName.ActivityType:
                return "activityType";
            default:
                throw new Error(
                    `Unknown or invalid metadata tag name: ${name}`
                );
        }
    }

    public static toPrefix(name: CodebookName): string {
        switch (name) {
            case CodebookName.Position:
                return "pos";
            case CodebookName.Calculation:
                return "calc";
            case CodebookName.Quantity:
                return "qty";
            case CodebookName.State:
                return "state";
            case CodebookName.Content:
                return "cnt";
            case CodebookName.Command:
                return "cmd";
            case CodebookName.Type:
                return "type";
            case CodebookName.Detail:
                return "detail";
            case CodebookName.FunctionalServices:
                return "funct.svc";
            case CodebookName.MaintenanceCategory:
                return "maint.cat";
            case CodebookName.ActivityType:
                return "act.type";
            default:
                throw new Error(
                    "Unknown or invalid metadata tag name: " + name
                );
        }
    }
}

File: cpp/src/dnv/vista/sdk/CodebookName.cpp
/**
 * @file CodebookName.cpp
 * @brief Implementation of codebook name conversion functions
 */

#include "pch.h"

#include "dnv/vista/sdk/CodebookName.h"

namespace dnv::vista::sdk
{
	//=====================================================================
	// Constants
	//=====================================================================

	namespace
	{
		static constexpr std::string_view POSITION_PREFIX = "pos";
		static constexpr std::string_view QUANTITY_PREFIX = "qty";
		static constexpr std::string_view CALCULATION_PREFIX = "calc";
		static constexpr std::string_view STATE_PREFIX = "state";
		static constexpr std::string_view CONTENT_PREFIX = "cnt";
		static constexpr std::string_view COMMAND_PREFIX = "cmd";
		static constexpr std::string_view TYPE_PREFIX = "type";
		static constexpr std::string_view FUNCTIONAL_SERVICES_PREFIX = "funct.svc";
		static constexpr std::string_view MAINTENANCE_CATEGORY_PREFIX = "maint.cat";
		static constexpr std::string_view ACTIVITY_TYPE_PREFIX = "act.type";
		static constexpr std::string_view DETAIL_PREFIX = "detail";
	}

	//=====================================================================
	// Enum mapping tables
	//=====================================================================

	namespace
	{
		static const std::unordered_map<std::string_view, CodebookName> s_prefixMap{
			{ POSITION_PREFIX, CodebookName::Position },
			{ QUANTITY_PREFIX, CodebookName::Quantity },
			{ CALCULATION_PREFIX, CodebookName::Calculation },
			{ STATE_PREFIX, CodebookName::State },
			{ CONTENT_PREFIX, CodebookName::Content },
			{ COMMAND_PREFIX, CodebookName::Command },
			{ TYPE_PREFIX, CodebookName::Type },
			{ FUNCTIONAL_SERVICES_PREFIX, CodebookName::FunctionalServices },
			{ MAINTENANCE_CATEGORY_PREFIX, CodebookName::MaintenanceCategory },
			{ ACTIVITY_TYPE_PREFIX, CodebookName::ActivityType },
			{ DETAIL_PREFIX, CodebookName::Detail } };
	}

	//=====================================================================
	// CodebookNames class
	//=====================================================================

	//----------------------------------------------
	// Public static methods
	//----------------------------------------------

	CodebookName CodebookNames::fromPrefix( const std::string_view prefix )
	{
		if ( prefix.empty() ) [[unlikely]]
		{
			throw std::invalid_argument( "Prefix cannot be empty." );
		}

		const auto it = s_prefixMap.find( prefix );
		if ( it != s_prefixMap.end() ) [[likely]]
		{
			return it->second;
		}

		throw std::invalid_argument( "Unknown prefix: " + std::string( prefix ) );
	}

	std::string_view CodebookNames::toPrefix( CodebookName name )
	{
		switch ( name )
		{
			case CodebookName::Position:
				return POSITION_PREFIX;
			case CodebookName::Quantity:
				return QUANTITY_PREFIX;
			case CodebookName::Calculation:
				return CALCULATION_PREFIX;
			case CodebookName::State:
				return STATE_PREFIX;
			case CodebookName::Content:
				return CONTENT_PREFIX;
			case CodebookName::Command:
				return COMMAND_PREFIX;
			case CodebookName::Type:
				return TYPE_PREFIX;
			case CodebookName::FunctionalServices:
				return FUNCTIONAL_SERVICES_PREFIX;
			case CodebookName::MaintenanceCategory:
				return MAINTENANCE_CATEGORY_PREFIX;
			case CodebookName::ActivityType:
				return ACTIVITY_TYPE_PREFIX;
			case CodebookName::Detail:
				return DETAIL_PREFIX;
			default:
			{
				throw std::invalid_argument( "Unknown codebook: " + std::to_string( static_cast<int>( name ) ) );
			}
		}
	}
}
